File: backend/app/services/pdf_extractor.py
import PyPDF2
import pdfplumber
from typing import List, Dict
import re

def word_count(text: str) -> int:
    """Count words in text"""
    return len(text.split())

def clean_text(text: str) -> str:
    """Clean and normalize text"""
    text = re.sub(r'\s+', ' ', text)  # Normalize whitespace
    text = re.sub(r'\n{3,}', '\n\n', text)  # Normalize line breaks
    return text.strip()

def is_meaningful_text(text: str) -> bool:
    """Check if text is meaningful (not just filler)"""
    trimmed = text.strip()
    
    # Skip very short or very long texts
    if len(trimmed) < 25 or len(trimmed) > 600:
        return False
    
    # Skip if it's just numbers or page numbers
    if re.match(r'^\d+$', trimmed) or re.match(r'^Page \d+', trimmed, re.IGNORECASE):
        return False
    
    # Skip common headers/footers
    header_footer_patterns = [
        r'^(Chapter|Section|Page \d+|\d+)$',
        r'^(Table of Contents|Index|Bibliography)$',
        r'^\d+$',
        r'^[A-Z\s]{1,5}$',
    ]
    
    for pattern in header_footer_patterns:
        if re.match(pattern, trimmed, re.IGNORECASE):
            return False
    
    # Must have some actual words
    words = [w for w in trimmed.split() if len(w) > 2]
    if len(words) < 4:
        return False
    
    # Skip if mostly special characters
    special_chars = len(re.findall(r'[^a-zA-Z0-9\s]', trimmed))
    if special_chars / len(trimmed) > 0.4:
        return False
    
    # Check for meaningful words
    meaningful_words = ['the', 'and', 'for', 'was', 'are', 'with', 'this', 'that', 'from', 'have']
    has_meaningful = any(word in trimmed.lower() for word in meaningful_words)
    
    return has_meaningful or len(words) >= 8

def is_quotation(text: str) -> bool:
    """Check if text looks like a quotation"""
    trimmed = text.strip()
    
    # Has quotation marks
    has_quotes = bool(re.search(r'["\'«»"]', trimmed))
    
    # Starts with capital letter
    starts_with_capital = bool(re.match(r'^[A-Z]', trimmed))
    
    # Has complete sentences
    has_complete_sentence = bool(re.search(r'[.!?]$', trimmed))
    
    # Contains dialogue indicators
    has_dialogue = bool(re.search(r'(said|says|told|asked|replied|exclaimed|whispered|shouted)', trimmed, re.IGNORECASE))
    
    # Is a statement
    sentences = [s.strip() for s in re.split(r'[.!?]', trimmed) if len(s.strip()) > 10]
    is_statement = len(sentences) >= 1
    
    return (has_quotes and has_complete_sentence) or \
           (starts_with_capital and is_statement and not has_dialogue and word_count(trimmed) >= 10)

def is_verse(text: str) -> bool:
    """Check if text looks like a verse"""
    trimmed = text.strip()
    
    # Numbered verse
    if re.match(r'^\d+[\.\)]\s+[A-Z]', trimmed):
        return True
    
    # Poetic structure
    lines = [l.strip() for l in trimmed.split('\n') if l.strip()]
    if 2 <= len(lines) <= 8:
        avg_length = sum(len(l) for l in lines) / len(lines)
        if 20 < avg_length < 100:
            capitalized = sum(1 for l in lines if re.match(r'^[A-Z]', l))
            if capitalized >= len(lines) * 0.7:
                return True
    
    # Short meaningful statement
    if 30 <= len(trimmed) <= 200 and \
       re.match(r'^[A-Z]', trimmed) and \
       5 <= word_count(trimmed) <= 40:
        sentence_count = len(re.findall(r'[.!?]', trimmed))
        if sentence_count <= 2:
            return True
    
    return False

async def extract_text_from_pdf(pdf_path: str) -> List[str]:
    """Extract text from PDF file, returning list of pages"""
    try:
        pages = []
        
        # Try using pdfplumber first (better text extraction)
        try:
            with pdfplumber.open(pdf_path) as pdf:
                for page in pdf.pages:
                    text = page.extract_text()
                    if text and text.strip():
                        pages.append(text.strip())
        except:
            # Fallback to PyPDF2
            with open(pdf_path, 'rb') as file:
                pdf_reader = PyPDF2.PdfReader(file)
                for page_num, page in enumerate(pdf_reader.pages):
                    text = page.extract_text()
                    if text and text.strip():
                        pages.append(text.strip())
        
        return pages if pages else [""]
    except Exception as e:
        raise Exception(f"Failed to extract text from PDF: {str(e)}")

def extract_items_from_text(pages: List[str]) -> List[Dict]:
    """Extract quotations, verses, and code from text pages"""
    items = []
    position = 0
    
    for page_index, page_text in enumerate(pages):
        page_number = page_index + 1
        cleaned_text = clean_text(page_text)
        
        # 1. Extract proper quotations (with quotation marks)
        quote_patterns = [
            r'"([^"]{30,500})"',
            r"'([^']{30,500})'",
            r'«([^»]{30,500})»',
            r'"([^"\n]{30,400})"',
        ]
        
        for pattern in quote_patterns:
            for match in re.finditer(pattern, cleaned_text, re.MULTILINE):
                content = clean_text(match.group(1) or match.group(0))
                if is_meaningful_text(content) and is_quotation(content):
                    # Check for duplicates
                    is_duplicate = any(
                        item['content'][:50].lower() == content[:50].lower()
                        for item in items
                    )
                    if not is_duplicate:
                        items.append({
                            'type': 'quote',
                            'content': content,
                            'pageNumber': page_number,
                            'position': position
                        })
                        position += 1
        
        # 2. Extract numbered verses
        verse_pattern = r'^\s*(\d+)[\.\)]\s+([^\n]{25,300})'
        for match in re.finditer(verse_pattern, cleaned_text, re.MULTILINE):
            content = clean_text(match.group(2))
            if is_meaningful_text(content):
                is_duplicate = any(
                    item['content'][:40].lower() == content[:40].lower()
                    for item in items
                )
                if not is_duplicate:
                    items.append({
                        'type': 'verse',
                        'content': content,
                        'pageNumber': page_number,
                        'position': position
                    })
                    position += 1
        
        # 3. Extract poetic verses
        lines = [l.strip() for l in cleaned_text.split('\n') if l.strip()]
        
        i = 0
        while i < len(lines) - 1:
            group = []
            j = i
            
            while j < len(lines) and len(group) < 6:
                line = lines[j]
                if 20 <= len(line) <= 150 and re.match(r'^[A-Z]', line):
                    group.append(line)
                    j += 1
                else:
                    break
            
            if 2 <= len(group) <= 6:
                verse_content = ' '.join(group)
                if is_verse(verse_content) and is_meaningful_text(verse_content):
                    is_duplicate = any(
                        item['content'][:40].lower() == verse_content[:40].lower()
                        for item in items
                    )
                    if not is_duplicate:
                        items.append({
                            'type': 'verse',
                            'content': verse_content,
                            'pageNumber': page_number,
                            'position': position
                        })
                        position += 1
                    i = j - 1
                    continue
            
            i += 1
        
        # 4. Extract meaningful standalone statements
        paragraphs = [clean_text(p) for p in cleaned_text.split('\n\n+')]
        
        for para in paragraphs:
            if not is_meaningful_text(para):
                continue
            
            word_cnt = word_count(para)
            sentence_count = len(re.findall(r'[.!?]', para))
            
            if (sentence_count >= 1 and sentence_count <= 3 and
                10 <= word_cnt <= 50 and
                re.match(r'^[A-Z]', para) and
                re.search(r'[.!?]$', para.strip())):
                
                # Check if already captured
                already_captured = any(
                    item['content'][:30] in para or para[:30] in item['content']
                    for item in items
                )
                
                if not already_captured:
                    item_type = 'quote' if is_quotation(para) else 'verse'
                    items.append({
                        'type': item_type,
                        'content': para,
                        'pageNumber': page_number,
                        'position': position
                    })
                    position += 1
        
        # 5. Extract code blocks
        code_patterns = [
            r'```[\s\S]{20,500}?```',
            r'<code>[\s\S]{20,500}?</code>',
        ]
        
        for pattern in code_patterns:
            for match in re.finditer(pattern, cleaned_text, re.IGNORECASE):
                content = clean_text(match.group(0))
                if 20 < len(content) < 1000:
                    items.append({
                        'type': 'code',
                        'content': content,
                        'pageNumber': page_number,
                        'position': position
                    })
                    position += 1
    
    # Final deduplication
    unique_items = []
    seen_content = set()
    
    for item in items:
        normalized = item['content'].lower()[:60]
        
        if normalized in seen_content:
            continue
        
        if not is_meaningful_text(item['content']):
            continue
        
        # Check for near-duplicates
        is_duplicate = any(
            seen[:50] == normalized[:50] for seen in seen_content
        )
        
        if not is_duplicate:
            seen_content.add(normalized)
            unique_items.append(item)
    
    # Sort by page number and position
    unique_items.sort(key=lambda x: (x['pageNumber'], x.get('position', 0)))
    
    # Limit to 300 high-quality items
    return unique_items[:300]

File: lib/pdf-extractor.ts
import pdf from 'pdf-parse'

interface ExtractedItem {
  type: 'verse' | 'quote' | 'code'
  content: string
  pageNumber: number
  position: number
}

export async function extractTextFromPDF(buffer: Buffer): Promise<string[]> {
  try {
    const data = await pdf(buffer)
    const pages: string[] = []
    
    // pdf-parse gives us full text, so we'll split by common page breaks
    const fullText = data.text
    
    // Try to split by page breaks or approximate by character count
    const totalPages = data.numpages || 1
    
    if (totalPages === 1) {
      // Single page or couldn't detect pages
      pages.push(fullText)
    } else {
      // Try to find page break patterns or split evenly
      const textLength = fullText.length
      const charsPerPage = Math.ceil(textLength / totalPages)
      
      for (let i = 0; i < totalPages; i++) {
        const start = i * charsPerPage
        const end = Math.min((i + 1) * charsPerPage, textLength)
        pages.push(fullText.slice(start, end))
      }
    }
    
    return pages.filter(page => page.trim().length > 0)
  } catch (error) {
    console.error('Error extracting text from PDF:', error)
    throw new Error('Failed to extract text from PDF')
  }
}

// Helper function to count words
function wordCount(text: string): number {
  return text.split(/\s+/).filter(w => w.length > 0).length
}

// Helper function to check if text is meaningful (not just filler)
function isMeaningfulText(text: string): boolean {
  const trimmed = text.trim()
  
  // Skip very short or very long texts
  if (trimmed.length < 25 || trimmed.length > 600) {
    return false
  }
  
  // Skip if it's just numbers or page numbers
  if (/^\d+$/.test(trimmed) || /^Page \d+/i.test(trimmed)) {
    return false
  }
  
  // Skip common headers/footers
  const headerFooterPatterns = [
    /^(Chapter|Section|Page \d+|\d+)$/i,
    /^(Table of Contents|Index|Bibliography)$/i,
    /^\d+$/,
    /^[A-Z\s]{1,5}$/, // All caps short text (likely headers)
  ]
  
  if (headerFooterPatterns.some(pattern => pattern.test(trimmed))) {
    return false
  }
  
  // Must have some actual words (not just punctuation/special chars)
  const wordCount = trimmed.split(/\s+/).filter(w => w.length > 2).length
  if (wordCount < 4) {
    return false
  }
  
  // Skip if it's mostly special characters or URLs
  const specialCharRatio = (trimmed.match(/[^a-zA-Z0-9\s]/g) || []).length / trimmed.length
  if (specialCharRatio > 0.4) {
    return false
  }
  
  // Check if it has meaningful content (contains common words)
  const meaningfulWords = ['the', 'and', 'for', 'was', 'are', 'with', 'this', 'that', 'from', 'have']
  const hasMeaningfulWords = meaningfulWords.some(word => 
    new RegExp(`\\b${word}\\b`, 'i').test(trimmed)
  )
  
  return hasMeaningfulWords || wordCount >= 8
}

// Helper function to check if text looks like a quotation
function isQuotation(text: string): boolean {
  const trimmed = text.trim()
  
  // Has quotation marks
  const hasQuotes = /["'«»"]/.test(trimmed)
  
  // Starts with capital letter
  const startsWithCapital = /^[A-Z]/.test(trimmed)
  
  // Has complete sentences (ends with punctuation)
  const hasCompleteSentence = /[.!?]$/.test(trimmed)
  
  // Contains dialogue indicators
  const hasDialogue = /(said|says|told|asked|replied|exclaimed|whispered|shouted)/i.test(trimmed)
  
  // Is a statement (not a question or command fragment)
  const isStatement = trimmed.split(/[.!?]/).filter(s => s.trim().length > 10).length >= 1
  
  // Must have quotes OR be a meaningful statement that sounds quotable
  return (hasQuotes && hasCompleteSentence) || 
         (startsWithCapital && isStatement && !hasDialogue && wordCount(trimmed) >= 10)
}

// Helper function to check if text looks like a verse
function isVerse(text: string): boolean {
  const trimmed = text.trim()
  
  // Numbered verse (e.g., "1. In the beginning...")
  if (/^\d+[\.\)]\s+[A-Z]/.test(trimmed)) {
    return true
  }
  
  // Poetic structure (short lines, often capitalized)
  const lines = trimmed.split('\n').filter(l => l.trim().length > 0)
  if (lines.length >= 2 && lines.length <= 8) {
    const avgLineLength = lines.reduce((sum, line) => sum + line.trim().length, 0) / lines.length
    if (avgLineLength > 20 && avgLineLength < 100) {
      // Check if multiple lines start with capital
      const capitalizedLines = lines.filter(line => /^[A-Z]/.test(line.trim())).length
      if (capitalizedLines >= lines.length * 0.7) {
        return true
      }
    }
  }
  
  // Short, meaningful statement that sounds like a verse or aphorism
  if (trimmed.length >= 30 && trimmed.length <= 200 && 
      /^[A-Z]/.test(trimmed) && 
      wordCount(trimmed) >= 5 && 
      wordCount(trimmed) <= 40) {
    const sentenceCount = (trimmed.match(/[.!?]/g) || []).length
    if (sentenceCount <= 2) {
      return true
    }
  }
  
  return false
}

// Helper to clean and normalize text
function cleanText(text: string): string {
  return text
    .replace(/\s+/g, ' ') // Normalize whitespace
    .replace(/\n{3,}/g, '\n\n') // Normalize line breaks
    .trim()
}

export function extractItemsFromText(pages: string[]): ExtractedItem[] {
  const items: ExtractedItem[] = []
  let position = 0

  pages.forEach((pageText, pageIndex) => {
    const pageNumber = pageIndex + 1
    const cleanedPageText = cleanText(pageText)
    
    // 1. Extract proper quotations (with quotation marks) - HIGHEST PRIORITY
    const quotePatterns = [
      /"([^"]{30,500})"/g, // Double quotes - meaningful quotes
      /'([^']{30,500})'/g, // Single quotes
      /«([^»]{30,500})»/g, // French quotes
      /"([^"\n]{30,400})"/gm, // Multi-line quotes
    ]
    
    quotePatterns.forEach((pattern) => {
      let match
      while ((match = pattern.exec(cleanedPageText)) !== null) {
        const content = cleanText(match[1] || match[0])
        if (isMeaningfulText(content) && isQuotation(content)) {
          const isDuplicate = items.some((item) => {
            const similarity = item.content.slice(0, 50).toLowerCase() === content.slice(0, 50).toLowerCase()
            return similarity
          })
          if (!isDuplicate) {
            items.push({
              type: 'quote',
              content,
              pageNumber,
              position: position++,
            })
          }
        }
      }
    })

    // 2. Extract numbered verses (e.g., "1. Verse text" or "1) Verse text")
    const verseNumberPattern = /^\s*(\d+)[\.\)]\s+([^\n]{25,300})/gm
    let verseMatch
    while ((verseMatch = verseNumberPattern.exec(cleanedPageText)) !== null) {
      const content = cleanText(verseMatch[2])
      if (isMeaningfulText(content)) {
        const isDuplicate = items.some((item) => {
          return item.content.slice(0, 40).toLowerCase() === content.slice(0, 40).toLowerCase()
        })
        if (!isDuplicate) {
          items.push({
            type: 'verse',
            content,
            pageNumber,
            position: position++,
          })
        }
      }
    }

    // 3. Extract poetic verses (multiple short lines, often capitalized)
    const lines = cleanedPageText.split('\n').map(l => l.trim()).filter(l => l.length > 0)
    
    // Look for groups of 2-6 consecutive lines that form a verse
    for (let i = 0; i < lines.length - 1; i++) {
      const group: string[] = []
      let j = i
      
      // Collect consecutive short lines
      while (j < lines.length && group.length < 6) {
        const line = lines[j]
        if (line.length >= 20 && line.length <= 150 && /^[A-Z]/.test(line)) {
          group.push(line)
          j++
        } else {
          break
        }
      }
      
      // If we found 2-6 lines, check if they form a verse
      if (group.length >= 2 && group.length <= 6) {
        const verseContent = group.join(' ')
        if (isVerse(verseContent) && isMeaningfulText(verseContent)) {
          const isDuplicate = items.some((item) => {
            return item.content.slice(0, 40).toLowerCase() === verseContent.slice(0, 40).toLowerCase()
          })
          if (!isDuplicate) {
            items.push({
              type: 'verse',
              content: verseContent,
              pageNumber,
              position: position++,
            })
            i = j - 1 // Skip processed lines
            continue
          }
        }
      }
    }

    // 4. Extract meaningful standalone statements (short, quotable paragraphs)
    const paragraphs = cleanedPageText.split(/\n\n+/).map(p => cleanText(p))
    
    paragraphs.forEach((para) => {
      if (!isMeaningfulText(para)) return
      
      const wordCnt = wordCount(para)
      const sentenceCount = (para.match(/[.!?]/g) || []).length
      
      // Look for quotable statements: 1-3 sentences, 10-50 words, starts with capital
      if (
        sentenceCount >= 1 && 
        sentenceCount <= 3 &&
        wordCnt >= 10 && 
        wordCnt <= 50 &&
        /^[A-Z]/.test(para) &&
        /[.!?]$/.test(para.trim())
      ) {
        // Check if it's already captured as a quote
        const alreadyCaptured = items.some((item) => {
          return item.content.includes(para.slice(0, 30)) || para.includes(item.content.slice(0, 30))
        })
        
        if (!alreadyCaptured) {
          // Prefer as quote if it sounds quotable, otherwise as verse
          const type: 'quote' | 'verse' = isQuotation(para) ? 'quote' : 'verse'
          
          items.push({
            type,
            content: para,
            pageNumber,
            position: position++,
          })
        }
      }
    })

    // 5. Extract code blocks (if any)
    const codePatterns = [
      /```[\s\S]{20,500}?```/g,
      /<code>[\s\S]{20,500}?<\/code>/gi,
    ]
    
    codePatterns.forEach((pattern) => {
      let codeMatch
      while ((codeMatch = pattern.exec(cleanedPageText)) !== null) {
        const content = cleanText(codeMatch[0])
        if (content.length > 20 && content.length < 1000) {
          items.push({
            type: 'code',
            content,
            pageNumber,
            position: position++,
          })
        }
      }
    })
  })

  // Final deduplication and quality filtering
  const uniqueItems: ExtractedItem[] = []
  const seenContent = new Set<string>()
  
  items.forEach((item) => {
    const normalizedContent = item.content.toLowerCase().slice(0, 60)
    
    // Skip if we've seen similar content
    if (seenContent.has(normalizedContent)) {
      return
    }
    
    // Additional quality check
    if (!isMeaningfulText(item.content)) {
      return
    }
    
    // Check for near-duplicates
    const isNearDuplicate = Array.from(seenContent).some(seen => {
      const similarity = seen.slice(0, 50) === normalizedContent.slice(0, 50)
      return similarity
    })
    
    if (!isNearDuplicate) {
      seenContent.add(normalizedContent)
      uniqueItems.push(item)
    }
  })

  // Sort by page number and position, then limit to reasonable number
  uniqueItems.sort((a, b) => {
    if (a.pageNumber !== b.pageNumber) {
      return a.pageNumber - b.pageNumber
    }
    return (a.position || 0) - (b.position || 0)
  })

  // Limit to max 300 high-quality items per book
  return uniqueItems.slice(0, 300)
}

